File: 207-course-schedule/course-schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        mapper = {}
        visited = set()

        for i in range(numCourses):
            mapper[i] = []

        for i in prerequisites:
            mapper[i[0]].append(i[1])

        def dfs(i):
            if i in visited:
                return False
            if len(mapper[i]) == 0:
                return True
            visited.add(i)
            for j in mapper[i]:
                if not dfs(j):
                    return False
            visited.remove(i)
            mapper[i] = []
            return True

        for i in range(numCourses):
            if dfs(i) == False: 
                return False
        return True
            

File: 207-course-schedule/course-schedule.java
import java.util.Vector;

class Solution {
    HashMap<Integer, Vector<Integer>> mapper = new HashMap<Integer, Vector<Integer>>();
    HashSet<Integer> visited = new HashSet<Integer>();

    boolean dfs(int i){
        if (visited.contains(i)) return false;
        if (mapper.getOrDefault(i, new Vector<Integer>()).size() == 0) return true;
        visited.add(i);
        for(int j: mapper.get(i))
            if (!dfs(j)) return false;
        visited.remove(i);
        mapper.get(i).removeAllElements();
        return true;
    }

    public boolean canFinish(int numCourses, int[][] prerequisites) {
        for(int i[]: prerequisites){
            Vector<Integer> temp = mapper.getOrDefault(i[0], new Vector<Integer>());
            temp.add(i[1]);
            mapper.put(i[0], temp);
        }
            
        
        for(int i = 0; i < numCourses; i++)
            if (!dfs(i)) return false;

        return true;        
    }
}